### DISCUSSION TITLE
Charsets and box drawing in WordPerfect 6.x

### DISCUSSION THREAD

**Discussion by dialognewmedia-mws** [dosemu2/dosemu2#1583] [Q&A]
What a great program DOSEMU2 is compared to the original! I've completely integrated it into my cross-device tmux setup for using WordPerfect 6.2 in terminal mode.

One issue I have noticed that was not present in the original DOSEMU is that the internal charsets seem to be performing differently.

By default, in WP6, the box drawing characters are incorrect. Menus and dialog boxes are not rendered with proper characters:
![photo_2022-01-18_16-55-50](https://user-images.githubusercontent.com/3652947/150043533-908b6c4c-2557-45a5-a975-e68783d5aae5.jpg)
 
But if you change $_external_char_set = "iso8859-1" then it fixes many of the issues. Menus look how they are supposed to, however dialog boxes are still not quite right:

![photo_2022-01-18_16-55-47](https://user-images.githubusercontent.com/3652947/150043734-06b3f8ce-3c3e-45e5-a3ba-ba8af24fbae1.jpg)

**Comment by stsp** [comment]
Thanks, there was a regression which
I now fixed. Please try again. It should
work w/o altering $_external_char_set.

**Reply by stsp** [reply]
You can use issues tracker for bugs.

**Reply by dialognewmedia-mws** [reply]
Thanks. I just tried building from devel branch but the instructions don't include info on how to install fdpp inside of the dosemu2 build.

Here's the last several lines of my ./default-configure run:

```
./configure --disable-option-checking --sysconfdir=/etc --with-fdtarball=/home/mws/Temp/dosemu2/dosemu-freedos-1.0-bin.tgz --enable-plugins=libao,gpm,alsa,ladspa,extra_charsets,term,X,Xkmaps,sdl,midimisc,fluidsynth,munt,ieee1284,charsets,json,slirp,console,modemu,dosdrv,doscmd,periph,debugger,fdpp,
configure: loading site script /usr/share/site/x86_64-unknown-linux-gnu
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking for FDPP... no
configure: error: fdpp installation not found or too old
Configuration for fdpp failed, disabling
configure: Compiling without debug info...
checking for __attribute__((force_align_arg_pointer))... supported
configure: Compiling with x86 emulator
configure: Compiling with default target CPU...
configure: Compiling as PIE for x86_64...
configure: optimization set to -O2
configure: error: fdpp not available!
```

I tried compiling fdpp inside of a subdirectory of dosemu2 but it doesn't seem to work. I then tried default-configure with the --disable-fdpp flag but that doesn't seem to work on the make step.

```
fatfs.c:124:37: error: ‘FDP_IDX’ undeclared here (not in a function); did you mean ‘FDP_D’?
  124 | static const uint64_t FDP_D = (1 << FDP_IDX);
      |                                     ^~~~~~~
      |                                     FDP_D
fatfs.c:124:23: warning: ‘FDP_D’ defined but not used [-Wunused-const-variable=]
  124 | static const uint64_t FDP_D = (1 << FDP_IDX);
      |                       ^~~~~
make[2]: *** [/home/mws/Temp/dosemu2/src/Makefile.common.post:89: fatfs.o] Error 1
make[2]: Leaving directory '/home/mws/Temp/dosemu2/src/base/misc'
make[1]: *** [/home/mws/Temp/dosemu2/src/arch/linux/Makefile.main:104: base/misc] Error 2
make[1]: Leaving directory '/home/mws/Temp/dosemu2/src'
make: *** [Makefile:30: default] Error 2

```

**Reply by stsp** [reply]
Compilation error is now fixed.
It is however a very bad idea to disable fdpp.
You can install it from our PPA or COPR,
or build from sources:
https://github.com/dosemu2/fdpp
Not sure if there is something specific here
to mention in a docs.

**Reply by dialognewmedia-mws** [reply]
Yes I built fdpp from source. That worked. I already had fdpp installed from packages though. How do I instruct the dosemu2 build to find it? Do I need a prefix or what? There's nothing in the documentation about where to put fdpp in relation to dosemu2 source.

**Reply by stsp** [reply]
> There's nothing in the documentation about where to put fdpp in relation to dosemu2 source.

But documentation says this:
```
building and installing

Just run make.
For an out-of-tree build, run <path_to_fdpp>/configure before running make.
After compiling, run sudo make install to install.
```
I was coviding, so couldn't even check
your claim, but its not missing in a doc. :)

**Comment by stsp** [comment]
Maybe you didnt install fdpp-devel?
Anyway, please just install dosemu2
from packages. It will pull all the deps.

**Comment by stsp** [comment]
There is nothing in documentation
because "sudo make install" should work.

**Comment by stsp** [comment]
All works?

**Reply by dialognewmedia-mws** [reply]
Thanks. I did get it to compile after doing ./configure in the fdpp directory. Unfortunately, now it can't run anything for me in terminal mode. Even well-behaved stuff like FreeDOS Edit. 

Now, dosemu.bin just locks up the virtual term and I have to kill the process externally. It runs character output stuff fine.

**Reply by stsp** [reply]
I am quite sure you now get a
mess of things installed from packages
and from sources, and you probably are
also running dosemu.bin directly from a
build dir...
I just suggest to uninstall everything and
use only binary packages.

**Reply by stsp** [reply]
You can use "sudo make uninstall" in
both dosemu2 and fdpp to uninstall the
source-based installation.

**Reply by dialognewmedia-mws** [reply]
It looks like there was some mixing of the binaries and the compiled stuff. I just uninstalled the packages and recompiled and it's working now and looks great. Thanks!
